### ISSUE TITLE
High Latency + Slow to reach peak bandwidth saturation with UCX-py 

### ISSUE THREAD

**Issue by Akshay-Venkatesh** [rapidsai/ucx-py#73]
```
akvenkatesh@prm-dgx-06:~/ucx-py$ python3 benchmarks/lat-bw.py -a -b -s X.X.X.X -p 13337
[1553779319.122772] [prm-dgx-06:72934:0] parser.c:1369 UCX WARN unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Size (bytes) Latency (us) BW (GB/s) Issue (us)
1 172.27 0.00
2 131.92 0.00
4 157.35 0.00
8 171.79 0.00
16 169.85 0.00
32 168.93 0.00
64 165.72 0.00
128 163.10 0.00
256 159.05 0.00
512 134.66 0.00
1024 145.15 0.01
2048 160.13 0.01
4096 162.51 0.03
8192 188.56 0.04
16384 206.67 0.08
32768 1642.30 0.02
65536 1750.97 0.04
131072 1841.53 0.07
262144 1935.85 0.14
524288 2091.98 0.25
1048576 2465.08 0.43
2097152 2855.87 0.73
4194304 4133.87 1.01
 
# without blind recvs
 
akvenkatesh@prm-dgx-06:~/ucx-py$ python3 benchmarks/lat-bw.py -a -s X.X.X.X -p 13337
[1553779473.298945] [prm-dgx-06:73294:0] parser.c:1369 UCX WARN unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Size (bytes) Latency (us) BW (GB/s) Issue (us)
1 135.65 0.00
2 149.53 0.00
4 139.20 0.00
8 150.10 0.00
16 131.43 0.00
32 127.36 0.00
64 117.91 0.00
128 108.97 0.00
256 129.06 0.00
512 127.63 0.00
1024 155.97 0.01
2048 130.15 0.02
4096 132.82 0.03
8192 122.11 0.07
16384 160.63 0.10
32768 1732.05 0.02
65536 1717.52 0.04
131072 1814.29 0.07
262144 1835.79 0.14
524288 2162.78 0.24
1048576 2418.00 0.43
2097152 2949.80 0.71
4194304 4134.40 1.01
```

Slow to reach peak bandwidth (compare 10MB and 100MB):
```
akvenkatesh@dgx15:~/akvenkatesh/ucx-py$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_TLS=rc,cuda_copy UCX_NET_DEVICES=mlx5_0:1 CUDA_VISIBLE_DEVICES=1 python3 benchmarks/recv-into-client.py -s X.X.X.X -p 13337 -o cupy
/home/nfs/akvenkatesh/akvenkatesh/py/lib/python3.7/site-packages/dask/config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
/home/nfs/akvenkatesh/akvenkatesh/distributed/distributed/config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)
[1553886254.375036] [dgx15:4210 :0]         parser.c:1369 UCX  WARN  unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Roundtrip benchmark
-------------------
n_iter   | 10
n_bytes  | 10.00 MB
recv     | recv_into
object   | cupy
inc      | False

===================
7.40 GB / s
===================

akvenkatesh@dgx15:~/akvenkatesh/ucx-py$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_TLS=rc,cuda_copy UCX_NET_DEVICES=mlx5_0:1 CUDA_VISIBLE_DEVICES=1 python3 benchmarks/recv-into-client.py -s X.X.X.X -p 13337 -o cupy -n 100Mb
/home/nfs/akvenkatesh/akvenkatesh/py/lib/python3.7/site-packages/dask/config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
/home/nfs/akvenkatesh/akvenkatesh/distributed/distributed/config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)
[1553887030.152355] [dgx15:4624 :0]         parser.c:1369 UCX  WARN  unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Roundtrip benchmark
-------------------
n_iter   | 10
n_bytes  | 100.00 MB
recv     | recv_into
object   | cupy
inc      | False

===================
10.30 GB / s
===================
```

**Comment by Akshay-Venkatesh** [comment]
Early performance analysis is indicating slowness in detection of read events being the culprit. Some details here:

```
$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy python3 benchmarks/lat-bw.py -a 
Size (bytes)    Latency (us)    BW (GB/s)       Send Issue (us) Recv Issue (us) Send Comp (us)  Recv Comp (us)
2               153.93          0.00            77.34           41.40           1.11            186.19
SEND STATS
2       16.347494992342863
RECV STATS
2       4.390153017911044
HANDLE_MSG STATS
-1      11.627999219027432
ACTIVITY STATS
-1      81.38841476993284
```
```
$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy python3 benchmarks/lat-bw.py -a 
[1555374790.767392] [prm-dgx-04:27695:0]         parser.c:1437 UCX  WARN  unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Size (bytes)    Latency (us)    BW (GB/s)       Send Issue (us) Recv Issue (us) Send Comp (us)  Recv Comp (us)
1024            148.40          0.01            72.41           38.64           1.05            183.04
SEND STATS
1024    15.993768518621271
RECV STATS
1024    3.85306098244407
HANDLE_MSG STATS
-1      11.148994619196111
ACTIVITY STATS
-1      92.65736393306567
```
```
$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy python3 benchmarks/lat-bw.py -a 
[1555374821.411977] [prm-dgx-04:27820:0]         parser.c:1437 UCX  WARN  unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Size (bytes)    Latency (us)    BW (GB/s)       Send Issue (us) Recv Issue (us) Send Comp (us)  Recv Comp (us)
32768           237.21          0.14            59.49           45.43           138.25          229.55
SEND STATS
32768   14.312267303466797
RECV STATS
32768   5.110393870960582
HANDLE_MSG STATS
-1      10.147853331132369
ACTIVITY STATS
-1      143.58939546527284
```
```
$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy python3 benchmarks/lat-bw.py -a 
[1555374924.880918] [prm-dgx-04:28225:0]         parser.c:1437 UCX  WARN  unused env variables: UCX_HOME, UCX_PATH, UCX_PY_CUDA_PATH,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
Size (bytes)    Latency (us)    BW (GB/s)       Send Issue (us) Recv Issue (us) Send Comp (us)  Recv Comp (us)
8388608         1313.23         6.39            106.01          81.78           1107.50         1327.85
SEND STATS
8388608 18.673376603560015
RECV STATS
8388608 8.2618539983576
HANDLE_MSG STATS
-1      15.219666741111062
ACTIVITY STATS
-1      230.85928546081792
```
https://gist.githubusercontent.com/Akshay-Venkatesh/368bcfdcf1da00d3b50270f39c2daf12/raw/4ab3b1020807bbe222a0a986c6fa7eee825b6d71/server-hist.log
https://gist.githubusercontent.com/Akshay-Venkatesh/368bcfdcf1da00d3b50270f39c2daf12/raw/4ab3b1020807bbe222a0a986c6fa7eee825b6d71/client-hist.log

**Comment by Akshay-Venkatesh** [comment]
- I've been able to narrow down on the culprit high latency with ucxpy

- for short messages: https://gist.github.com/Akshay-Venkatesh/368bcfdcf1da00d3b50270f39c2daf12/revisions#diff-21189b76dce45b69408917b664a5412f

 - for large messages: https://gist.github.com/Akshay-Venkatesh/368bcfdcf1da00d3b50270f39c2daf12/revisions#diff-ea22c5ce84c7a98c47eb6ec833c01aab

 - this two links contains rows of the form `[0.0, x, y, z..]` where each row represents a message request
 - 0.0 represents the time when it was created
 - x, y, z represents the time at which the message was checked for completion in reader_loop (the callback called by python event loop when it notices an event on the fd returned by ucx)
 - what we see for the small messages (recv reqs) is that it's generally of the form `[0.0, ~200]`
 - basically, this explains why we have huge latencies in the short message range
 - for large messages this is of the form `[0.0, ~200, ~2000]`
 - so there's an event read (for initial control message) at ~200, and then one at ~2000 (for fin and thus request completion)

**Comment by Akshay-Venkatesh** [comment]
Small update on this - @quasiben's uvloop suggestion:

Using uvloop is showing some benefits with almost zero code changes:
```shell
# default asyncio

akvenkatesh@prm-dgx-03:~/ucx-py$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_NET_DEVICES=mlx5_0:1 UCX_TLS=rc,cuda_copy CUDA_VISIBLE_DEVICES=1 python3 benchmarks/recv-into-client.py -o cupy -s IP -p 13337
Roundtrip benchmark
-------------------
n_iter   | 10
n_bytes  | 10.00 MB
recv     | recv_into
object   | cupy
inc      | False

===================
5.31 GB / s
===================

# with uvloop

akvenkatesh@prm-dgx-03:~/ucx-py$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_NET_DEVICES=mlx5_0:1 UCX_TLS=rc,cuda_copy CUDA_VISIBLE_DEVICES=1 python3 benchmarks/recv-into-client.py -o cupy -s IP -p 13337 -u
Roundtrip benchmark
-------------------
n_iter   | 10
n_bytes  | 10.00 MB
recv     | recv_into
object   | cupy
inc      | False

===================
6.11 GB / s
===================
```

Changes needed:
```diff
$ git diff benchmarks/recv-into-client.py 
diff --git a/benchmarks/recv-into-client.py b/benchmarks/recv-into-client.py
index bba9eb8..a01adf5 100644
--- a/benchmarks/recv-into-client.py
+++ b/benchmarks/recv-into-client.py
@@ -9,6 +9,7 @@ from distributed.utils import format_bytes, parse_bytes
 from time import perf_counter as clock
 
 import ucp
+import uvloop
 
 
 ucp.init()
@@ -31,6 +32,8 @@ def parse_args(args=None):
                         help="In-memory array type.")
     parser.add_argument("-v", "--verbose", default=False, action="store_true",
                         help="Whether to print timings per iteration.")
+    parser.add_argument("-u", "--uvloop", default=False, action="store_true",
+                        help="Whether or not to use uvloop")
     parser.add_argument("-i", "--inc", default=False, action="store_true",
                         help="Whether to increment the array each iteration.")
 
@@ -144,4 +147,7 @@ async def main(args=None):
 
 
 if __name__ == '__main__':
+    args = parse_args()
+    if args.uvloop:
+        uvloop.install()
     asyncio.run(main())

```

**Comment by Akshay-Venkatesh** [comment]
Differences in terms of latency:
```shell
# without uvloop 

$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_NET_DEVICES=mlx5_0:1 UCX_TLS=rc,cuda_copy CUDA_
VISIBLE_DEVICES=0 python3 benchmarks/lat-bw.py -a
Size (bytes)    Latency (us)    BW (GB/s)       Issue (us)
1               127.90          0.00            58.35           27.53           1.06            166.69
2               125.67          0.00            49.88           24.17           0.84            174.46
4               104.44          0.00            41.20           19.67           0.68            145.78
8               113.98          0.00            51.55           25.18           0.89            148.31
16              133.52          0.00            58.05           27.25           0.99            178.59
32              130.27          0.00            58.69           27.46           0.98            171.24
64              127.61          0.00            55.81           26.18           1.00            170.18
128             129.48          0.00            60.13           28.38           1.07            167.14
256             127.30          0.00            58.60           27.76           1.01            164.99
512             132.47          0.00            60.89           29.38           1.04            171.36
1024            122.50          0.01            57.49           28.61           0.95            155.70
2048            121.17          0.02            67.59           27.49           1.04            144.13
4096            112.61          0.04            70.24           29.24           1.48            121.94
8192            105.97          0.08            63.20           25.11           1.01            120.70
16384           106.47          0.15            62.33           24.12           0.98            123.59
32768           212.51          0.15            51.61           37.70           124.50          208.88
65536           213.18          0.31            55.28           40.47           127.93          200.23
131072          251.69          0.52            57.70           42.35           159.29          241.32
262144          301.16          0.87            63.36           46.50           193.51          296.04
524288          326.32          1.61            66.11           48.54           219.51          315.45
1048576         390.22          2.69            67.24           49.62           279.74          380.80
2097152         639.87          3.28            90.16           59.94           435.05          690.43
4194304         814.66          5.15            79.57           57.53           644.31          844.17

# with uvloop 

$ UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy UCX_NET_DEVICES=mlx5_0:1 UCX_TLS=rc,cuda_copy CUDA_VISIBLE_DEVICES=0 python3 benchmarks/lat-bw.py -a
Size (bytes)    Latency (us)    BW (GB/s)       Issue (us)
1               96.44           0.00            56.68           27.37           0.95            105.18
2               90.34           0.00            54.07           26.21           0.88            96.96
4               82.70           0.00            47.02           22.62           0.70            92.82
8               67.31           0.00            39.85           19.21           0.67            73.01
16              90.95           0.00            54.59           26.81           0.84            97.08
32              76.09           0.00            44.54           21.88           0.66            83.01
64              85.40           0.00            54.55           26.91           0.88            85.79
128             73.08           0.00            43.31           20.96           0.69            79.22
256             62.37           0.00            36.64           18.05           0.55            67.77
512             74.07           0.01            39.36           19.39           0.58            86.97
1024            77.43           0.01            41.63           20.40           0.63            90.30
2048            68.42           0.03            50.95           21.05           0.70            62.30
4096            82.66           0.05            64.11           26.70           0.91            71.29
8192            64.72           0.13            50.46           20.59           0.82            55.76
16384           66.50           0.25            52.05           20.80           0.71            57.60
32768           127.55          0.26            42.36           30.89           67.03           112.49
65536           138.30          0.47            49.66           36.12           73.43           114.60
131072          171.45          0.76            54.46           39.33           95.09           151.02
262144          196.76          1.33            55.70           40.50           116.99          177.29
524288          234.88          2.23            61.33           44.94           153.13          206.98
1048576         296.46          3.54            60.82           44.18           207.09          277.31
2097152         478.29          4.38            73.80           49.88           317.96          510.60
4194304         639.91          6.55            66.69           48.55           519.43          641.19
```

**Comment by mrocklin** [comment]
That's really cool to see.

**Comment by Akshay-Venkatesh** [comment]
@mrocklin Another finding - I see that from the time the UCX file descriptor is notified of a communication completion event to the point when asyncio loop.add_reader method sees is separated by about 120 microseconds (90 us with uvloop). If we take out memory allocation out of the picture (supposing we get an efficient memory manager), then this would be the main overhead.

**Comment by mrocklin** [comment]
Strictly from an ETL standpoint I'm not too troubled by this.  

We care far more about bandwidth than latency, and are often sending around messages that are in the 100MB range.  The tables above tell me that I should be quite happy.

However, expanding to other applications, like parameter servers in deep learning, yes, I do become concerned, but that's fairly far down the line for me personally.

From a RAPIDS standpoint, handling the latency here is a lowish priority.

**Comment by mrocklin** [comment]
Still happy to engage here though (I suspect that even if I don't care as much currently, that you do).  

At one point I think that you benchmarked asyncio and found that it added a few microseconds.   Reading your comment above more closely it sounds like we might be limited by the file descriptor/non-blocking socket behavior.  Is that correct?  Is this bound by the OS or Python?

One thing that might be interesting to try would be to run things under PyPy, a jit-ed Python interpreter.  I think I've gotten Torando roundtrips down to 60us roundtrip when doing experiments before.

Also, I recall that Torando had a few event loops.  EPoll, KQueue, perhaps others.  I wonder if there might be some other OS technology that we can rely on.

**Comment by mrocklin** [comment]
This is also the kind of thing that @pitrou might know about.  

@pitrou if you have any interest I think that you might enjoy speaking with @Akshay-Venkatesh at some point.  Akshay works on [OpenUCX](http://www.openucx.org/), a high performance networking library that we're connecting to Python (for Dask at first, but the work is general purpose).  Akshay is mostly a C/C++ performance-oriented developer who is now dealing with Python async frameworks, event loops, networking, and so on.  I suspect that you're busy these days, but you might enjoy the conversation.

**Comment by Akshay-Venkatesh** [comment]
> At one point I think that you benchmarked asyncio and found that it added a few microseconds. Reading > your comment above more closely it sounds like we might be limited by the file descriptor/non-blocking > socket behavior. Is that correct? Is this bound by the OS or Python?

I think this is an asyncio loop issue. uvloop alleviates the issue a little bit. Having said this, does use of PyPy as opposed to CPython still make a difference?

As long as there's a more efficient event loop that lets us read file descriptor notifications quicker (or express a way to unblock a blocked coroutine), I'm happy to try.

**Comment by pitrou** [comment]
If this involves asyncio and uvloop then @1st1 may be interested to hear about it.

**Comment by pitrou** [comment]
(asyncio involves several layers and has its IO loop implemented in Python, so perhaps a 100 µs latency is expected, though unfortunate)

**Comment by quasiben** [comment]
I think we can close this issue as it relates to the previous version of ucx-py.  For those interested we are profiling/benchmarking and making good progress with workloads which use both cupy and cuDF
